File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;
class Solution {
    public int[] solution(int[] numbers) {
        
        Stack<Integer> stack = new Stack<>();
        
        stack.push(numbers[numbers.length-1]);
        
        int[] result = new int[numbers.length];
        
        result[numbers.length-1] = -1;
        
        for (int i=numbers.length-2; i>=0; i--) {
            if (numbers[i] >= stack.get(0)) {
                result[i] = -1;
                stack = new Stack<>();
                stack.push(numbers[i]);
            } else {
                for (int j=stack.size()-1; j>=0; j--) {
                    int st = stack.get(j);
                    if (st > numbers[i]) {
                        result[i] = st;
                        stack.push(numbers[i]);
                        break;
                    } else {
                        stack.pop();
                    }
                }
            }            
        }
        return result;
    }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    
    result = [-1 for i in range(len(numbers))]
    
    stack = [numbers[len(numbers)-1]]
    
    for i in range(len(numbers)-2, -1, -1):
        if stack[0] > numbers[i]:
            for j in range(len(stack)-1, -1, -1):
                if stack[j] > numbers[i]:
                    result[i] = stack[j]
                    break
                else:
                    stack.pop()
            stack.append(numbers[i])
            
        else:
            result[i] = -1
            stack = [numbers[i]]
             
    answer = []
    return result